### PR TITLE
feat: add `http.request`

### LINKED ISSUE #1849: Support for the http module
### What version of Bun is running?

0.5.0

### What platform is your computer?

Darwin 22.2.0 arm64 arm

### What steps can reproduce the bug?

1) Run the following (or any script that uses the http module from nodeJS): 
```
const http = require('http');

const options = {
  host: 'www.google.com',
  path: '/',
  method: 'GET'
};

const req = http.request(options, (res) => {
  console.log(`statusCode: ${res.statusCode}`);

  res.on('data', (d) => {
    process.stdout.write(d);
  });
});

req.on('error', (error) => {
  console.error(error);
});

req.end();
```

2) Get this error: ```TypeError: undefined is not a function (near '...http.request...')```

### What is the expected behavior?

This returns the expected response (the result of a request to google) when I run it with node.

### What do you see instead?

TypeError: undefined is not a function (near '...http.request...')

### Additional information

I'm guessing this is happening because bun doesn't have support for the HTTP module yet. I'm just making an issue to track that.

### PR DESCRIPTION
**Pull request by ThatOneBro**
Still needs more tests and more cleanup. Going to start porting node tests for this and `node:net` next probably. Discovering small divergences between Node and Bun as testing corner cases, so this is really important to make sure we get closer to 100% Node compatibility

Closes #1849

### CODE REVIEW

File: src/bun.js/http.exports.js [hunk 1]
```diff
@@ -1,12 +1,199 @@
...
+const ObjectAssign = Object.assign;
+const ObjectPrototypeHasOwnProperty = Object.prototype.hasOwnProperty;
+
+const INVALID_PATH_REGEX = /[^\u0021-\u00ff]/;
+const NODE_HTTP_WARNING =
+  "WARN: Agent is mostly unused in Bun's implementation of http. If you see strange behavior, this is probably the cause.";
+
+var _globalAgent;
+
+class FakeSocket {
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: the class name should be `Socket` incase someone checks that instead of using `instanceof`

File: src/bun.js/http.exports.js [hunk 2]
```diff
@@ -1,12 +1,199 @@
...
+  keepSocketAlive() {
+    debug(`${NODE_HTTP_WARNING}\n`, "WARN: Agent.keepSocketAlive is a no-op");
+  }
+
+  reuseSocket() {
+    debug(`${NODE_HTTP_WARNING}\n`, "WARN: Agent.reuseSocket is a no-op");
+  }
+
+  destroy() {
+    debug(`${NODE_HTTP_WARNING}\n`, "WARN: Agent.destroy is a no-op");
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js:189]: Did you check these functions don't expect to return this?

File: src/bun.js/http.exports.js [hunk 3]
```diff
@@ -128,23 +315,28 @@ function destroyBodyStreamNT(bodyStream) {
   bodyStream.destroy();
 }
 export class IncomingMessage extends Readable {
-  constructor(req) {
+  #options;
+  constructor(req, options = { type: "request" }) {
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: do we need to preserve the original options object entirely? or just `type`?

File: src/bun.js/http.exports.js [hunk 4]
```diff
@@ -292,6 +486,209 @@ export class IncomingMessage extends Readable {
...
+function write_(msg, chunk, encoding, callback, fromEnd) {
+  if (typeof callback !== "function") callback = nop;
+
+  let len;
+  if (chunk === null) {
+    // throw new ERR_STREAM_NULL_VALUES();
+    throw new Error("ERR_STREAM_NULL_VALUES");
+  } else if (typeof chunk === "string") {
+    len = Buffer.byteLength(chunk, encoding);
+  } else if (isUint8Array(chunk)) {
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: Buffer.byteLength on a Uint8Array returns the byte length of the Uint8Array, meaning this check is unnecessary

File: src/bun.js/http.exports.js [hunk 5]
```diff
@@ -292,6 +486,209 @@ export class IncomingMessage extends Readable {
...
+  //   process.nextTick(connectionCorkNT, msg.socket);
+  // }
+
+  return true;
+}
+
+const kClearTimeout = Symbol("kClearTimeout");
+
+export class OutgoingMessage extends Writable {
+  #headers = new Headers();
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: Can we make this a lazy getter

File: src/bun.js/http.exports.js [hunk 6]
```diff
@@ -292,6 +486,209 @@ export class IncomingMessage extends Readable {
...
+
+  #finished = false;
+  [kEndCalled] = false;
+
+  #socket;
+  #timeoutTimer = null;
+
+  // So that headers are accessible outside of `OutgoingMessage`
+  get _headers() {
+    return this.#headers;
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: Are you sure it's supposed to return a Headers object? If this is an internal-only symbol ("protected") then can you make it a Symbol so that it doesn't appear as enumerable

File: src/bun.js/http.exports.js [hunk 7]
```diff
@@ -292,6 +486,209 @@ export class IncomingMessage extends Readable {
...
+
+  set useChunkedEncodingByDefault(value) {
+    // throw new Error('not implemented');
+  }
+
+  get socket() {
+    var _socket = this.#socket;
+    if (_socket) return _socket;
+
+    this.#socket = _socket = new EventEmitter();
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: Can this use the FakeSocket abstraction?

File: src/bun.js/http.exports.js [hunk 8]
```diff
@@ -292,6 +486,209 @@ export class IncomingMessage extends Readable {
...
+  }
+
+  [kClearTimeout]() {
+    if (this.#timeoutTimer) {
+      clearTimeout(this.#timeoutTimer);
+      this.#timeoutTimer = null;
+    }
+  }
+
+  setTimeout(msecs, callback) {
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js:648]: can you add a TODO: about AbortSignal here?

File: src/bun.js/http.exports.js [hunk 9]
```diff
@@ -517,6 +900,391 @@ export class ServerResponse extends Writable {
...
+      callback();
+      return;
+    }
+    this.#body = body + chunks.join();
+    callback();
+  }
+
+  _final(callback) {
+    this.#finished = true;
+    this.#fetchRequest = globalThis
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js]: can you set `fetch` to a global at the top that way we don't go through the globalThis proxy each time

```js
const fetch = Bun.fetch;
```

File: src/bun.js/http.exports.js [hunk 10]
```diff
@@ -517,6 +900,391 @@ export class ServerResponse extends Writable {
...
+    callback();
+
+    // TODO: Clear timeout here
+  }
+
+  get aborted() {
+    return this.#aborted;
+  }
+
+  abort() {
```
Comment (Jarred-Sumner) [inline src/bun.js/http.exports.js:974]:
```
// TODO: use AbortSignal from fetch
```

File: src/bun.js/http.exports.js (hunk 2 above)
Comment (ThatOneBro) [reply src/bun.js/http.exports.js:189]: Yes none of the Agent functions actually return this, I think because they are mostly used internally from within `http.request` or `http.Server`, so they aren't really expected to be chained

File: src/bun.js/http.exports.js (hunk 6 above)
Comment (ThatOneBro) [reply src/bun.js/http.exports.js]: Yeah I think I was referencing the partial implementation from #1627, I don't think this is actually being used now. Removed it

**Comment by Jarred-Sumner** [comment]
Thank you!

**Comment by Electroid** [comment]
Fn9YOxSaYAA-3rJ
